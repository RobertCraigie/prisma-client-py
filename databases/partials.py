from prisma.models import Post

Post.create_partial('PostOnlyPublished', include={'id', 'published'})
