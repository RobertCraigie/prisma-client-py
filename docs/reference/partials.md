# Partial Types

Prisma Client Python exposes an interface for creating partial models at generation time based off of schema-defined models, see [Getting started with partial types](../getting_started/partial-types.md) for more information.

Each model defined in the schema can have partial types generated based off of it. The API looks like this:
```py
class Model:
  @staticmethod
  def create_partial(
      name: str,
      include: Optional[Iterable['{{ model }}Keys']] = None,
      exclude: Optional[Iterable['{{ model }}Keys']] = None,
      required: Optional[Iterable['{{ model }}Keys']] = None,
      optional: Optional[Iterable['{{ model }}Keys']] = None,
      relations: Optional[Mapping['{{ model }}RelationalKeys', str]] = None,
      exclude_relational_fields: bool = False,
  ) -> None:
      ...
```
where `{{ model }}Keys` and `{{ model }}RelationalKeys` are [Literal](https://docs.python.org/3/library/typing.html#typing.Literal) types containing all the fields of the model and all the relational fields of the model respectively.

For example:

```prisma
model User {
  id      String   @default(cuid()) @id
  name    String
  posts   Post[]
  email   String?
}
```

```py
UserKeys = Literal['id', 'name', 'posts', 'email']
UserRelationalKeys = Literal['posts']
```

## Reference


`name`
: The name given to the generated partial type, must be unique.

`include`
: An iterable of field names that will be available in the generated model.
  Cannot be used at the same time as `exclude`.

`exclude`
: An iterable of field names that will *not* be available in the generated model.
  Cannot be used at the same time as `include`.

`required`
: An iterable of field names that will *not* be marked as optional in the generated model.

`optional`
: An iterable of field names that will be marked as optional in the generated model.

`relations`
: A mapping of relational field names to a custom partial model. (Cannot be used at the same time as `exclude_relational_fields`)

`exclude_relational_fields`
: A boolean indicating whether or not relational fields should be included in the generated model.

## Examples

All examples use the following models

```prisma
model User {
  id      String   @default(cuid()) @id
  name    String
  profile Profile?
  email   String?
}

model Profile {
  id       Int    @id @default(autoincrement())
  user     User   @relation(fields: [user_id], references: [id])
  user_id  String
  bio      String
}
```

```py
class User:
  id: str
  name: str
  profile: Optional['Profile']
  email: Optional[str]

class Profile:
  id: int
  user: Optional['User']
  user_id: str
  bio: str
```


### Include

```py
from prisma.models import User
User.create_partial('UserOnlyName', include={'name'})
User.create_partial('UserOnlyNameAndEmail', include={'name', 'email'})
```

```py
class UserOnlyName:
  name: str

class UserOnlyNameAndEmail:
  name: str
  email: Optional[str]
```


### Exclude

```py
from prisma.models import User
User.create_partial('UserWithoutProfile', exclude=['profile'])
```

```py
class UserWithoutProfile:
  id: str
  name: str
  email: Optional[str]
```

### Optional

```py
from prisma.models import User
User.create_partial('UserOptionalName', optional={'name'})
```

```py
class UserOptionalName:
  id: str
  name: Optional[str]
  profile: Optional['Profile']
  email: Optional[str]
```

### Required

```py
from prisma.models import User
User.create_partial('UserRequiredEmail', required={'email'})
```

```py
class UserRequiredEmail:
  id: str
  name: str
  profile: Optional['Profile']
  email: str
```

### Relations

```py
from prisma.models import User, Profile
Profile.create_partial('ProfileWithoutUser', exclude={'user'})
User.create_partial('UserCustomProfile', relations={'profile': 'ProfileWithoutUser'})
```

```py
class UserCustomProfile:
  id: str
  name: str
  profile: Optional['partials.ProfileWithoutUser']
  email: Optionalstr]
```

### Excluding Relational Fields

```py
from prisma.models import User
User.create_partial('UserWithoutRelations', exclude_relational_fields=True)
```

```py
class UserWithoutRelations:
  id: str
  name: str
  email: Optional[str]
```
