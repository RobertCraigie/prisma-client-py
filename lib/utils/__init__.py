from ._utils import flatten as flatten
