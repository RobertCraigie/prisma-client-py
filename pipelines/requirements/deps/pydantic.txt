pydantic==2.3.0
