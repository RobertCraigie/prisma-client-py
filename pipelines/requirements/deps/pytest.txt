pytest==7.4.4
