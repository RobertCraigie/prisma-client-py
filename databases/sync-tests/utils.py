from __future__ import annotations

import os
from pydantic import BaseModel

from .._types import DatabaseMapping
from .._compat import LiteralString


CURRENT_DATABASE = os.environ['PRISMA_DATABASE']


class RawQueries(BaseModel):
    """Raw queries defined globally so they can be easily referenced in test functions"""

    count_posts: LiteralString

    find_user_by_id: LiteralString
    find_user_by_id_limit_1: LiteralString

    find_post_by_id: LiteralString
    find_posts_not_published: LiteralString

    select_unknown_table: LiteralString
    update_unique_post_title: LiteralString
    update_unique_post_new_title: LiteralString
    test_query_raw_no_result: LiteralString
    test_execute_raw_no_result: LiteralString


_mysql_queries = RawQueries(
    count_posts="""
        SELECT COUNT(*) as count
        FROM Post
    """,
    find_post_by_id="""
        SELECT *
        FROM Post
        WHERE id = ?
    """,
    find_user_by_id="""
        SELECT *
        FROM User
        WHERE User.id = ?
    """,
    find_user_by_id_limit_1="""
        SELECT *
        FROM User
        WHERE User.id = ?
        LIMIT 1
    """,
    select_unknown_table="""
        SELECT *
        FROM bad_table;
    """,
    find_posts_not_published="""
        SELECT id, published
        FROM Post
        WHERE published = false
    """,
    test_query_raw_no_result="""
        SELECT *
        FROM Post
        WHERE id = 'sdldsd'
    """,
    update_unique_post_title="""
        UPDATE Post
        SET title = 'My edited title'
        WHERE id = ?
    """,
    update_unique_post_new_title="""
        UPDATE Post
        SET title = 'My new title'
        WHERE id = ?
    """,
    test_execute_raw_no_result="""
        UPDATE Post
        SET title = 'updated title'
        WHERE id = 'sdldsd'
    """,
)

_postgresql_queries = RawQueries(
    count_posts="""
        SELECT COUNT(*) as count
        FROM "Post"
    """,
    find_post_by_id="""
        SELECT *
        FROM "Post"
        WHERE id = $1
    """,
    find_user_by_id="""
        SELECT *
        FROM "User"
        WHERE "User".id = $1
    """,
    find_user_by_id_limit_1="""
        SELECT *
        FROM "User"
        WHERE "User".id = $1
        LIMIT 1
    """,
    select_unknown_table="""
        SELECT *
        FROM bad_table;
    """,
    find_posts_not_published="""
        SELECT id, published
        FROM "Post"
        WHERE published = false
    """,
    test_query_raw_no_result="""
        SELECT *
        FROM "Post"
        WHERE id = 'sdldsd'
    """,
    update_unique_post_title="""
        UPDATE "Post"
        SET title = 'My edited title'
        WHERE id = $1
    """,
    update_unique_post_new_title="""
        UPDATE "Post"
        SET title = 'My new title'
        WHERE id = $1
    """,
    test_execute_raw_no_result="""
        UPDATE "Post"
        SET title = 'updated title'
        WHERE id = 'sdldsd'
    """,
)

RAW_QUERIES_MAPPING: DatabaseMapping[RawQueries] = {
    'postgresql': _postgresql_queries,
    'cockroachdb': _postgresql_queries,
    'mysql': _mysql_queries,
    'mariadb': _mysql_queries,
    'sqlite': RawQueries(
        count_posts="""
            SELECT COUNT(*) as count
            FROM Post
        """,
        find_post_by_id="""
            SELECT *
            FROM Post
            WHERE id = ?
        """,
        find_user_by_id="""
            SELECT *
            FROM User
            WHERE User.id = ?
        """,
        find_user_by_id_limit_1="""
            SELECT *
            FROM User
            WHERE User.id = ?
            LIMIT 1
        """,
        select_unknown_table="""
            SELECT *
            FROM bad_table;
        """,
        find_posts_not_published="""
            SELECT id, published
            FROM Post
            WHERE published = false
        """,
        test_query_raw_no_result="""
            SELECT *
            FROM Post
            WHERE id = 'sdldsd'
        """,
        update_unique_post_title="""
            UPDATE Post
            SET title = 'My edited title'
            WHERE id = ?
        """,
        update_unique_post_new_title="""
            UPDATE Post
            SET title = 'My new title'
            WHERE id = ?
        """,
        test_execute_raw_no_result="""
            UPDATE Post
            SET title = 'updated title'
            WHERE id = 'sdldsd'
        """,
    ),
}
