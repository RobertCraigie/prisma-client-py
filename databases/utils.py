from __future__ import annotations

import os
from typing import Set
from pathlib import Path
from typing_extensions import Literal, get_args, override

from pydantic import BaseModel
from syrupy.location import PyTestLocation
from syrupy.extensions.amber import AmberSnapshotExtension

from ._types import DatabaseMapping
from ._compat import LiteralString

DatabaseFeature = Literal[
    'enum',
    'json',
    'date',
    'arrays',
    'array_push',
    'json_arrays',
    'raw_queries',
    'create_many_skip_duplicates',
    'transactions',
    'case_sensitivity',
    'full_text_search',
]


class DatabaseConfig(BaseModel):
    id: str
    name: str
    env_var: str
    bools_are_ints: bool
    autoincrement_id: str
    unsupported_features: Set[DatabaseFeature]
    default_date_func: str

    # TODO: run this under coverage
    def supports_feature(self, feature: DatabaseFeature) -> bool:  # pragma: no cover
        if feature not in get_args(DatabaseFeature):
            raise RuntimeError(f'Unknown feature: {feature}')

        return feature not in self.unsupported_features


# ------------------ Test helpers ------------------

from .constants import TESTS_DIR, SYNC_TESTS_DIR

SHARED_SNAPSHOTS_DIR = Path(__file__).parent.joinpath('__shared_snapshots__')
CURRENT_DATABASE = os.environ.get('PRISMA_DATABASE')


class AmberSharedExtension(AmberSnapshotExtension):
    """Syrupy extension that stores the snapshots in a parent __shared_snapshots__ dir"""

    @classmethod
    @override
    def dirname(cls, *, test_location: PyTestLocation) -> str:
        test_dir = Path(test_location.filepath).parent
        if test_dir.is_relative_to(SYNC_TESTS_DIR):
            rel_dir = test_dir.relative_to(SYNC_TESTS_DIR)
        else:
            rel_dir = test_dir.relative_to(TESTS_DIR)

        return str(SHARED_SNAPSHOTS_DIR.joinpath(rel_dir))


class RawQueries(BaseModel):
    """Raw queries defined globally so they can be easily referenced in test functions"""

    count_posts: LiteralString

    find_user_by_id: LiteralString
    find_user_by_id_limit_1: LiteralString

    find_post_by_id: LiteralString
    find_posts_not_published: LiteralString

    select_unknown_table: LiteralString
    update_unique_post_title: LiteralString
    update_unique_post_new_title: LiteralString
    test_query_raw_no_result: LiteralString
    test_execute_raw_no_result: LiteralString


_mysql_queries = RawQueries(
    count_posts="""
        SELECT COUNT(*) as count
        FROM Post
    """,
    find_post_by_id="""
        SELECT *
        FROM Post
        WHERE id = ?
    """,
    find_user_by_id="""
        SELECT *
        FROM User
        WHERE User.id = ?
    """,
    find_user_by_id_limit_1="""
        SELECT *
        FROM User
        WHERE User.id = ?
        LIMIT 1
    """,
    select_unknown_table="""
        SELECT *
        FROM bad_table;
    """,
    find_posts_not_published="""
        SELECT id, published
        FROM Post
        WHERE published = false
    """,
    test_query_raw_no_result="""
        SELECT *
        FROM Post
        WHERE id = 'sdldsd'
    """,
    update_unique_post_title="""
        UPDATE Post
        SET title = 'My edited title'
        WHERE id = ?
    """,
    update_unique_post_new_title="""
        UPDATE Post
        SET title = 'My new title'
        WHERE id = ?
    """,
    test_execute_raw_no_result="""
        UPDATE Post
        SET title = 'updated title'
        WHERE id = 'sdldsd'
    """,
)

_postgresql_queries = RawQueries(
    count_posts="""
        SELECT COUNT(*) as count
        FROM "Post"
    """,
    find_post_by_id="""
        SELECT *
        FROM "Post"
        WHERE id = $1
    """,
    find_user_by_id="""
        SELECT *
        FROM "User"
        WHERE "User".id = $1
    """,
    find_user_by_id_limit_1="""
        SELECT *
        FROM "User"
        WHERE "User".id = $1
        LIMIT 1
    """,
    select_unknown_table="""
        SELECT *
        FROM bad_table;
    """,
    find_posts_not_published="""
        SELECT id, published
        FROM "Post"
        WHERE published = false
    """,
    test_query_raw_no_result="""
        SELECT *
        FROM "Post"
        WHERE id = 'sdldsd'
    """,
    update_unique_post_title="""
        UPDATE "Post"
        SET title = 'My edited title'
        WHERE id = $1
    """,
    update_unique_post_new_title="""
        UPDATE "Post"
        SET title = 'My new title'
        WHERE id = $1
    """,
    test_execute_raw_no_result="""
        UPDATE "Post"
        SET title = 'updated title'
        WHERE id = 'sdldsd'
    """,
)

RAW_QUERIES_MAPPING: DatabaseMapping[RawQueries] = {
    'postgresql': _postgresql_queries,
    'cockroachdb': _postgresql_queries,
    'mysql': _mysql_queries,
    'mariadb': _mysql_queries,
    'sqlite': RawQueries(
        count_posts="""
            SELECT COUNT(*) as count
            FROM Post
        """,
        find_post_by_id="""
            SELECT *
            FROM Post
            WHERE id = ?
        """,
        find_user_by_id="""
            SELECT *
            FROM User
            WHERE User.id = ?
        """,
        find_user_by_id_limit_1="""
            SELECT *
            FROM User
            WHERE User.id = ?
            LIMIT 1
        """,
        select_unknown_table="""
            SELECT *
            FROM bad_table;
        """,
        find_posts_not_published="""
            SELECT id, published
            FROM Post
            WHERE published = false
        """,
        test_query_raw_no_result="""
            SELECT *
            FROM Post
            WHERE id = 'sdldsd'
        """,
        update_unique_post_title="""
            UPDATE Post
            SET title = 'My edited title'
            WHERE id = ?
        """,
        update_unique_post_new_title="""
            UPDATE Post
            SET title = 'My new title'
            WHERE id = ?
        """,
        test_execute_raw_no_result="""
            UPDATE Post
            SET title = 'updated title'
            WHERE id = 'sdldsd'
        """,
    ),
}
