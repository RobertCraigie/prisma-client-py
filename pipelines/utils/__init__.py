from ._lark import generate_parsers as generate_parsers
from .utils import *
