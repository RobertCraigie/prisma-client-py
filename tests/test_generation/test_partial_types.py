# pyright: reportUnusedFunction=false

import subprocess
from typing_extensions import Literal

import pytest
from ..utils import Testdir


SCHEMA = """
datasource db {{
  provider = "postgres"
  url      = env("DB_URL")
}}

generator db {{
  provider = "coverage run -m prisma"
  output = "{output}"
  {options}
}}

model Post {{
  id          String     @id @default(cuid())
  created_at  DateTime   @default(now())
  updated_at  DateTime   @updatedAt
  title       String
  published   Boolean
  desc        String?
  meta        Json?
  comments    Comment[]
  author_id   String
  author      User      @relation(fields: [author_id], references: [id])
  thumbnail   Bytes?
}}

model Comment {{
  id           String   @id @default(cuid())
  created_at   DateTime @default(now())
  updated_at   DateTime @updatedAt
  content      String
  post         Post?    @relation(fields: [post_id], references: [id])
  post_id      String?
}}

model User {{
  id           String   @id @default(cuid())
  created_at   DateTime @default(now())
  updated_at   DateTime @updatedAt
  name         String
  bytes        Bytes
  bytes_list   Bytes[]
  posts        Post[]
}}

model Foo {{
  id   String @id @default(cuid())
  text String
}}
"""


@pytest.mark.parametrize(
    'location,options',
    [
        ('prisma/partial_types.py', ''),
        (
            'prisma/partial_types.py',
            'partial_type_generator = "prisma/partial_types.py"',
        ),
        (
            'scripts/partials_generator.py',
            'partial_type_generator = "scripts.partials_generator"',
        ),
    ],
)
def test_partial_types(testdir: Testdir, location: str, options: str) -> None:
    """Grouped tests for partial types to improve test speed"""

    def tests() -> None:  # mark: filedef
        import sys
        import datetime
        from typing import Type, Dict, Iterator, Any, Tuple, Set, Optional
        from pydantic import BaseModel
        from prisma import Base64
        from prisma.partials import (  # type: ignore[attr-defined]
            # pyright: reportGeneralTypeIssues = false
            PostWithoutDesc,
            PostOptionalPublished,
            PostRequiredDesc,
            PostOnlyId,
            PostNoRelations,
            PostOptionalInclude,
            PostRequiredAuthor,
            PostModifiedAuthor,
            UserModifiedPosts,
            UserBytesList,
            PostNoRelationsAndExclude,
        )

        base_fields = {
            'id': True,
            'created_at': True,
            'updated_at': True,
            'title': True,
            'published': True,
            'desc': False,
            'meta': False,
            'comments': False,
            'author': False,
            'author_id': True,
            'thumbnail': False,
        }

        def common_entries(
            dct: Dict[str, Any], other: Dict[str, Any]
        ) -> Iterator[Tuple[str, Any, Any]]:
            for key in dct.keys() & other.keys():
                yield key, dct[key], other[key]

        def assert_expected(
            model: Type[BaseModel],
            fields: Dict[str, bool],
            removed: Optional[Set[str]],
        ) -> None:
            for field, required, info in common_entries(
                fields, model.__fields__
            ):
                assert info.name == field
                assert info.required == required

            if removed is None:
                removed = set()

            assert fields.keys() - model.__fields__.keys() == removed

        def test_without_desc() -> None:
            """Removing one field"""
            assert_expected(PostWithoutDesc, base_fields, {'desc'})

        def test_optional_published() -> None:
            """Making one field optional"""
            assert_expected(
                PostOptionalPublished,
                fields={**base_fields, 'published': False},
                removed=None,
            )

        def test_required_desc() -> None:
            """Making one field required"""
            assert_expected(
                PostRequiredDesc,
                fields={**base_fields, 'desc': True},
                removed=None,
            )

        def test_required_relational_author() -> None:
            """Making relational field required"""
            assert_expected(
                PostRequiredAuthor,
                fields={**base_fields, 'author': True},
                removed=None,
            )

        def test_only_id() -> None:
            """Removing all fields except from one"""
            assert_expected(
                PostOnlyId,
                fields=base_fields,
                removed={
                    'created_at',
                    'updated_at',
                    'title',
                    'published',
                    'desc',
                    'meta',
                    'comments',
                    'author',
                    'author_id',
                    'thumbnail',
                },
            )

        def test_optional_include() -> None:
            """Both including and making optional on the same field"""
            assert_expected(
                PostOptionalInclude,
                fields={**base_fields, 'title': False},
                removed={
                    'id',
                    'created_at',
                    'updated_at',
                    'published',
                    'desc',
                    'meta',
                    'comments',
                    'author',
                    'author_id',
                    'thumbnail',
                },
            )

        def test_modified_relational_type() -> None:
            """Changing one-to-one relational field type"""
            assert_expected(PostModifiedAuthor, base_fields, None)

            field = PostModifiedAuthor.__fields__['author']
            assert field.type_.__name__ == 'UserOnlyName'
            assert field.type_.__module__ == 'prisma.partials'

        def test_exclude_relations() -> None:
            """Removing all relational fields using `exclude_relations`"""
            assert_expected(
                PostNoRelations,
                base_fields,
                removed={
                    'author',
                    'comments',
                },
            )

        def test_exclude_relations_and_others() -> None:
            """Removing all relational fields using `exclude_relations` in combination with `exclude`"""
            assert_expected(
                PostNoRelationsAndExclude,
                base_fields,
                removed={
                    'title',
                    'author',
                    'comments',
                },
            )

        def test_modified_relational_list_type() -> None:
            """Changing one-to-many relation field type"""
            UserModifiedPosts(
                id='1',
                name='Robert',
                created_at=datetime.datetime.utcnow(),
                updated_at=datetime.datetime.utcnow(),
                posts=[PostOnlyId(id='2')],
            )
            field = UserModifiedPosts.__fields__['posts']
            assert field.type_.__name__ == 'PostOnlyId'
            assert field.type_.__module__ == 'prisma.partials'

            if sys.version_info >= (3, 7):
                assert field.outer_type_._name == 'List'
            else:
                assert field.outer_type_.__name__ == 'List'

            assert field.outer_type_.__module__ == 'typing'

        def test_bytes() -> None:
            """Ensure Base64 fields can be used"""
            # mock prisma behaviour
            model = UserBytesList.parse_obj(
                {
                    'bytes': str(Base64.encode(b'bar')),
                    'bytes_list': [
                        str(Base64.encode(b'foo')),
                        str(Base64.encode(b'baz')),
                    ],
                }
            )
            assert model.bytes == Base64.encode(b'bar')
            assert model.bytes_list == [
                Base64.encode(b'foo'),
                Base64.encode(b'baz'),
            ]

    def generator() -> None:  # mark: filedef
        from prisma.models import Post, User

        User.create_partial('UserOnlyName', include={'name'})

        Post.create_partial('PostWithoutDesc', exclude=['desc'])
        Post.create_partial('PostOptionalPublished', optional=['published'])
        Post.create_partial('PostRequiredDesc', required=['desc'])
        Post.create_partial('PostOnlyId', include={'id'})
        Post.create_partial(
            'PostOptionalInclude', include={'title'}, optional={'title'}
        )
        Post.create_partial('PostRequiredAuthor', required=['author'])
        Post.create_partial(
            'PostModifiedAuthor', relations={'author': 'UserOnlyName'}
        )
        Post.create_partial('PostNoRelations', exclude_relational_fields=True)
        Post.create_partial(
            'PostNoRelationsAndExclude',
            exclude={'title'},
            exclude_relational_fields=True,
        )

        User.create_partial(
            'UserModifiedPosts',
            exclude={'bytes', 'bytes_list'},  # type: ignore
            relations={'posts': 'PostOnlyId'},
        )
        User.create_partial('UserBytesList', include={'bytes', 'bytes_list'})  # type: ignore

    testdir.make_from_function(generator, name=location)
    testdir.generate(SCHEMA, options)
    testdir.make_from_function(tests)
    testdir.runpytest().assert_outcomes(passed=11)


@pytest.mark.parametrize(
    'argument', ['exclude', 'include', 'required', 'optional']
)
def test_partial_types_incorrect_key(
    testdir: Testdir,
    argument: Literal['exclude', 'include', 'required', 'optional'],
) -> None:
    """Invalid field name raises error"""

    def generator() -> None:  # mark: filedef
        from prisma.models import Post

        Post.create_partial('PostWithoutFoo', **{argument: ['foo']})  # type: ignore

    testdir.make_from_function(
        generator, name='prisma/partial_types.py', argument=argument
    )

    with pytest.raises(subprocess.CalledProcessError) as exc:
        testdir.generate(SCHEMA)

    assert 'foo is not a valid Post / PostWithoutFoo field' in str(
        exc.value.output
    )


def test_partial_types_same_required_and_optional(testdir: Testdir) -> None:
    """Making the same field required and optional raises an error"""

    def generator() -> None:  # mark: filedef
        from prisma.models import Post

        Post.create_partial(
            'PostPartial',
            required={'desc', 'published', 'title'},
            optional={
                'desc',
                'published',
            },
        )

    testdir.make_from_function(generator, name='prisma/partial_types.py')

    with pytest.raises(subprocess.CalledProcessError) as exc:
        testdir.generate(SCHEMA)

    assert 'Cannot make the same field(s) required and optional' in str(
        exc.value.output
    )


def test_partial_types_excluded_required(testdir: Testdir) -> None:
    """Excluding and requiring the same field raises an error"""

    def generator() -> None:  # mark: filedef
        from prisma.models import Post

        Post.create_partial(
            'PostPartial',
            exclude={'desc'},
            required={
                'desc',
            },
        )

    testdir.make_from_function(generator, name='prisma/partial_types.py')

    with pytest.raises(subprocess.CalledProcessError) as exc:
        testdir.generate(SCHEMA)

    assert 'desc is not a valid Post / PostPartial field' in str(
        exc.value.output
    )


def test_partial_type_generator_not_found(testdir: Testdir) -> None:
    """Unknown partial type generator option value raises an error"""
    with pytest.raises(subprocess.CalledProcessError) as exc:
        testdir.generate(SCHEMA, 'partial_type_generator = "foo.bar.baz"')

    output = str(exc.value.output)
    assert 'ValidationError' in output
    assert 'generator -> config -> partial_type_generator' in output
    assert 'Could not find a python file or module at foo.bar.baz' in output


def test_partial_type_generator_error_while_running(testdir: Testdir) -> None:
    """Exception ocurring while running partial type generator logs exception"""

    def generator() -> None:  # mark: filedef
        import foo  # type: ignore[import]

    testdir.make_from_function(generator, name='prisma/partial_types.py')

    with pytest.raises(subprocess.CalledProcessError) as exc:
        testdir.generate(SCHEMA)

    output = str(exc.value.output)
    assert 'prisma/partial_types.py' in output
    assert "No module named \\'foo\\'" in output
    assert (
        'An exception ocurred while running the partial type generator'
        in output
    )


def test_partial_type_already_created(testdir: Testdir) -> None:
    """Creating same partial type twice raises error"""

    def generator() -> None:  # mark: filedef
        from prisma.models import Post

        for _ in range(2):
            Post.create_partial(
                'PostPartial',
                exclude={'desc'},
            )

    testdir.make_from_function(generator, name='prisma/partial_types.py')

    with pytest.raises(subprocess.CalledProcessError) as exc:
        testdir.generate(SCHEMA)

    output = str(exc.value.output)
    assert 'prisma/partial_types.py' in output
    assert 'Partial type "PostPartial" has already been created.' in output
    assert (
        'An exception ocurred while running the partial type generator'
        in output
    )


def test_unknown_partial_type(testdir: Testdir) -> None:
    """Modifying relational field type to unknown partial type raises error"""

    def generator() -> None:  # mark: filedef
        from prisma.models import Post

        Post.create_partial('PostPartial', relations={'author': 'UnknownUser'})

    testdir.make_from_function(generator, name='prisma/partial_types.py')

    with pytest.raises(subprocess.CalledProcessError) as exc:
        testdir.generate(SCHEMA)

    output = str(exc.value.output)
    assert 'ValueError' in output
    assert 'prisma/partial_types.py' in output
    assert 'Unknown partial type: "UnknownUser"' in output
    assert (
        'An exception ocurred while running the partial type generator'
        in output
    )
    assert (
        'Did you remember to generate the UnknownUser type before this one?'
        in output
    )


def test_passing_type_for_excluded_field(testdir: Testdir) -> None:
    """Passing relational type for an excluded field raises error"""

    def generator() -> None:  # mark: filedef
        from prisma.models import Post, User

        User.create_partial('CustomUser')
        Post.create_partial(
            'PostPartial',
            exclude={'author'},
            relations={'author': 'CustomUser'},
        )

    testdir.make_from_function(generator, name='prisma/partial_types.py')

    with pytest.raises(subprocess.CalledProcessError) as exc:
        testdir.generate(SCHEMA)

    output = str(exc.value.output)
    assert 'ValueError' in output
    assert 'prisma/partial_types.py' in output
    assert 'author is not a valid Post / PostPartial field' in output
    assert (
        'An exception ocurred while running the partial type generator'
        in output
    )


def test_partial_type_types_non_relational(testdir: Testdir) -> None:
    """Passing non-relational field to relations raises an error"""

    def generator() -> None:  # mark: filedef
        from prisma.models import Post

        Post.create_partial('Placeholder')
        Post.create_partial(
            'PostPartial',
            relations={'published': 'Placeholder'},  # type: ignore[dict-item]
        )

    testdir.make_from_function(generator, name='prisma/partial_types.py')

    with pytest.raises(subprocess.CalledProcessError) as exc:
        testdir.generate(SCHEMA)

    output = str(exc.value.output)
    assert 'prisma/partial_types.py' in output
    assert 'prisma.errors.UnknownRelationalFieldError' in output
    assert (
        'An exception ocurred while running the partial type generator'
        in output
    )
    assert (
        'Field: "published" either does not exist or is not a relational field on the Post model'
        in output
    )


def test_partial_type_relations_no_relational_fields(testdir: Testdir) -> None:
    """Passing relations option to model with no relational fields raises an error"""

    def generator() -> None:  # mark: filedef
        from prisma.models import Foo  # type: ignore[attr-defined]

        Foo.create_partial('Placeholder', relations={'wow': 'Placeholder'})

    testdir.make_from_function(generator, name='prisma/partial_types.py')

    with pytest.raises(subprocess.CalledProcessError) as exc:
        testdir.generate(SCHEMA)

    output = exc.value.output.decode('utf-8')
    assert 'prisma/partial_types.py' in output
    assert 'ValueError' in output
    assert (
        'An exception ocurred while running the partial type generator'
        in output
    )
    assert 'Model: "Foo" has no relational fields.' in output


def test_exclude_relational_fields_and_relations_exclusive(
    testdir: Testdir,
) -> None:
    """exclude_relational_fields and relations cannot be passed at the same time"""

    def generator() -> None:  # mark: filedef
        from prisma.models import Post

        Post.create_partial(
            'Placeholder',
            relations={'author': 'Placeholder'},
            exclude_relational_fields=True,
        )

    testdir.make_from_function(generator, name='prisma/partial_types.py')

    with pytest.raises(subprocess.CalledProcessError) as exc:
        testdir.generate(SCHEMA)

    output = exc.value.output.decode('utf-8')
    assert 'prisma/partial_types.py' in output
    assert 'ValueError' in output
    assert (
        'An exception ocurred while running the partial type generator'
        in output
    )
    assert (
        'exclude_relational_fields and relations are mutually exclusive'
        in output
    )


def test_exclude_relational_fields_and_include_exclusive(
    testdir: Testdir,
) -> None:
    """exclude_relational_fields and include cannot be passed at the same time"""

    def generator() -> None:  # mark: filedef
        from prisma.models import Post

        Post.create_partial(
            'Placeholder',
            include={'author'},
            exclude_relational_fields=True,
        )

    testdir.make_from_function(generator, name='prisma/partial_types.py')

    with pytest.raises(subprocess.CalledProcessError) as exc:
        testdir.generate(SCHEMA)

    output = exc.value.output.decode('utf-8')
    assert 'prisma/partial_types.py' in output
    assert 'TypeError' in output
    assert (
        'An exception ocurred while running the partial type generator'
        in output
    )
    assert (
        'Include and exclude_relational_fields=True are mutually exclusive.'
        in output
    )
