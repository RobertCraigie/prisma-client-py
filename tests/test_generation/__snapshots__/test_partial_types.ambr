# name: test_partial_type_generator_not_found
  list([
    "  Value error, Could not find a python file or module at foo.bar.baz [type=value_error, input_value='foo.bar.baz', input_type=str]",
    '    For further information visit https://errors.pydantic.dev/2.2/v/value_error',
    '',
  ])
# ---
